pkg/agent/proxy/proxier.go

```diff
@@ -536,58 +544,70 @@ func (p *proxier) installServices() {
 					}
 					// If previous Service which has ClusterIP should be removed, remove ClusterIP routes.
 					if svcInfo.ClusterIP() != nil {
 						if err := p.routeClient.DeleteClusterIPRoute(pSvcInfo.ClusterIP()); err != nil {
-							klog.ErrorS(err, "Failed to remove ClusterIP Service routes", "Service", svcPortName)
+							klog.ErrorS(err, "Error when uninstalling ClusterIP route for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 							continue
 						}
 					}
 				}
 			}
 
 			// Install ClusterIP flows for the Service.
-			groupID := p.groupCounter.AllocateIfNotExist(svcPortName, internalPolicyLocal)
+			groupID, exists := p.groupCounter.Get(svcPortName, internalPolicyLocal)
+			if !exists {
+				klog.InfoS("Group for Service internalTrafficPolicy is not installed", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr, "internalTrafficPolicy", internalPolicyLocal)
+				continue
+			}
 			if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(affinityTimeout), externalPolicyLocal, corev1.ServiceTypeClusterIP); err != nil {
-				klog.Errorf("Error when installing Service flows: %v", err)
+				klog.ErrorS(err, "Error when installing ClusterIP flows for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 				continue
 			}
 
 			if p.proxyAll {
-				nGroupID := p.groupCounter.AllocateIfNotExist(svcPortName, externalPolicyLocal)
 				// Install ClusterIP route on Node so that ClusterIP can be accessed on Node. Every time a new ClusterIP
 				// is created, the routing target IP block will be recalculated for expansion to be able to route the new
 				// created ClusterIP. Deleting a ClusterIP will not shrink the target routing IP block. The Service CIDR
 				// can be finally calculated after creating enough ClusterIPs.
 				if err := p.routeClient.AddClusterIPRoute(svcInfo.ClusterIP()); err != nil {
-					klog.ErrorS(err, "Failed to install ClusterIP route of Service", "Service", svcPortName)
+					klog.ErrorS(err, "Error when installing ClusterIP route for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 					continue
 				}
 
 				// If previous Service is nil or NodePort flows and configurations of previous Service have been removed,
 				// install NodePort flows and configurations for current Service.
 				if svcInfo.NodePort() > 0 && (pSvcInfo == nil || needRemoval) {
-					if err := p.installNodePortService(nGroupID, uint16(svcInfo.NodePort()), svcInfo.OFProtocol, uint16(affinityTimeout), svcInfo.ExternalPolicyLocal()); err != nil {
-						klog.ErrorS(err, "Failed to install NodePort flows and configurations of Service", "Service", svcPortName)
+					groupID, exists = p.groupCounter.Get(svcPortName, externalPolicyLocal)
+					if !exists {
+						klog.InfoS("Group for Service externalTrafficPolicy is not installed", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr, "externalTrafficPolicy", externalPolicyLocal)
```

Comment:
Is this an expected scenario after this patch?

Reply:
No, after this patch, for every Service, the cached Endpoints will always be the same as those installed in OVS, case like #4681 shouldn't happen,  but we discussed this before, a `Get` method should be used here.

Reply:
I have no problem to use `Get` here, but the log should be Error to indicate this is not something we expect.

Reply:
Thanks, updated.